File: app/src/main/java/com/siziksu/services/ui/activity/WeatherServiceActivity.java
package com.siziksu.services.ui.activity;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import com.siziksu.services.R;
import com.siziksu.services.app.Constants;
import com.siziksu.services.commons.Commons;
import com.siziksu.services.commons.DeviceManager;
import com.siziksu.services.data.service.WeatherService;
import com.siziksu.services.data.service.binder.WeatherBinder;

public class WeatherServiceActivity extends AppCompatActivity
        implements View.OnClickListener, WeatherBinder.WeatherListener {

    private Button buttonStart;

    private boolean bound;
    private WeatherService service;
    private ServiceConnection serviceConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_CONNECTED);
            WeatherServiceActivity.this.service = ((WeatherBinder) service).getService();
            bound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_DISCONNECTED);
            service = null;
            bound = false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_one_button);
        ((TextView) findViewById(R.id.activityTitle)).setText(
                getIntent().getStringExtra(Constants.EXTRAS_TITLE));
        ((TextView) findViewById(R.id.activitySummary)).setText(
                getIntent().getStringExtra(Constants.EXTRAS_SUMMARY));
        buttonStart = findViewById(R.id.btnStartService);
        buttonStart.setOnClickListener(this);
        DeviceManager.init(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (DeviceManager.getInstance().isServiceRunning(Constants.TAG_WEATHER_SERVICE)) {
            buttonStart.setEnabled(false);
            Toast.makeText(this, Constants.SERVICE_RUNNING, Toast.LENGTH_SHORT).show();
        }
        bindService();
    }

    @Override
    protected void onPause() {
        super.onPause();
        unbindService();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btnStartService:
                buttonStart.setEnabled(false);
                startService();
                break;
        }
    }

    private void startService() {
        Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_STARTING);
        startService(new Intent(getBaseContext(), WeatherService.class));
        service.getForecast(WeatherServiceActivity.this,
                getString(R.string.weather_text_to_format));
    }

    private void bindService() {
        if (!bound) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_BINDING);
            bindService(new Intent(getBaseContext(), WeatherService.class), serviceConnection,
                    Context.BIND_AUTO_CREATE);
        }
    }

    private void unbindService() {
        if (bound) {
            bound = false;
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_UNBINDING);
            unbindService(serviceConnection);
        }
    }

    @Override
    public void updateForecast(String forecast) {
        Commons.log(Constants.TAG_WEATHER_SERVICE, forecast);
    }

    @Override
    public void onForecastError(Exception e) {
        e.printStackTrace();
    }
}

File: app/src/main/java/com/siziksu/services/ui/activity/WeatherServiceActivity.kt
package com.siziksu.services.ui.activity

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.siziksu.services.R
import com.siziksu.services.commons.Commons
import com.siziksu.services.commons.Constants
import com.siziksu.services.commons.DeviceManager
import com.siziksu.services.data.service.WeatherService
import com.siziksu.services.data.service.binder.WeatherBinder
import kotlinx.android.synthetic.main.activity_one_button.one_button_layout
import kotlinx.android.synthetic.main.section_single_button.btnStartService
import kotlinx.android.synthetic.main.section_title.activitySummary
import kotlinx.android.synthetic.main.section_title.activityTitle

class WeatherServiceActivity : AppCompatActivity(), WeatherBinder.Listener {

    private var bound: Boolean = false
    private var service: WeatherService? = null
    private val serviceConnection = object : ServiceConnection {

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_CONNECTED)
            this@WeatherServiceActivity.service = (service as WeatherBinder).service
            bound = true
        }

        override fun onServiceDisconnected(name: ComponentName) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_DISCONNECTED)
            service = null
            bound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_one_button)
        initializeViews()
    }

    override fun onResume() {
        super.onResume()
        if (DeviceManager.isServiceRunning(this, Constants.TAG_WEATHER_SERVICE)) {
            btnStartService.isEnabled = false
            Snackbar.make(one_button_layout, Constants.SERVICE_RUNNING, Snackbar.LENGTH_SHORT).show()
        }
        bindService()
    }

    override fun onPause() {
        super.onPause()
        unbindService()
    }

    override fun updateForecast(forecast: String) {
        Commons.log(Constants.TAG_WEATHER_SERVICE, forecast)
    }

    private fun initializeViews() {
        activityTitle.text = intent.getStringExtra(Constants.EXTRAS_TITLE)
        activitySummary.text = intent.getStringExtra(Constants.EXTRAS_SUMMARY)
        btnStartService.setOnClickListener {
            btnStartService.isEnabled = false
            startService()
        }
    }

    private fun startService() {
        Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_STARTING)
        startService(Intent(baseContext, WeatherService::class.java))
        service?.getForecast(this@WeatherServiceActivity, getString(R.string.weather_text_to_format))
    }

    private fun bindService() {
        if (!bound) {
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_BINDING)
            bindService(Intent(baseContext, WeatherService::class.java), serviceConnection, Context.BIND_AUTO_CREATE)
        }
    }

    private fun unbindService() {
        if (bound) {
            bound = false
            Commons.log(Constants.TAG_WEATHER_SERVICE, Constants.SERVICE_UNBINDING)
            unbindService(serviceConnection)
        }
    }
}
